spec/controllers/users/omniauth_callbacks_controller_spec.rb

```diff
@@ -0,0 +1,170 @@
+      role: "contester",
+      name: "Test user",
+      city: "Айтос"
+    }
+
+    User.delete_all
+  end
+
+  describe "#google_oauth2" do
+
+    context "user does not exist" do
+
+      it "should redirect to edit_user_path" do
+        post :google_oauth2
+        expect(response).to redirect_to(edit_user_path(User.last))
+      end
+
+      it "should successfully create a user" do
+        expect {
+          post :google_oauth2
+        }.to change{ User.count }.by(1)
+
+        user = User.last
+        mock = mock_auth_hash()
+        expect(user.email).to eq(mock[:info][:email])
+        expect(user.name).to eq(mock[:info][:name])
+        expect(user.uid).to eq(mock[:uid])
+        expect(user.provider).to eq(mock[:provider])
+        expect(user.login).to eq(mock[:uid])
+        expect(user.role).to eq(User::CONTESTER)
+      end
+
+
+      it "should set a :notice and :error flash" do
+        post :google_oauth2
+        expect(flash[:notice]).to be_present
+        expect(flash[:notice]).to eq "Успешно оторизиран чрез Google профил."
+        expect(flash[:error]).to be_present
+        expect(flash[:error]).to eq "Данните в профила Ви не са пълни. Моле попълнете липсващите данни."
+      end
+    end
+
+    context "user exists with the same uid and provider" do
+
+
+      it "should not create a user" do
+        User.create(@user_with_uid_provider)
+        expect {
+          post :google_oauth2
+        }.to_not change{ User.count }
+      end
+
+      it "should set a :notice flash" do
+        User.create(@user_with_uid_provider)
+        post :google_oauth2
+        expect(flash[:notice]).to be_present
+        expect(flash[:notice]).to eq "Успешно оторизиран чрез Google профил."
+      end
+
+    end
+
+    context "user exists with the same email" do
+
+      it "should not create a user" do
+        User.create(@user_with_email)
+        expect {
+          post :google_oauth2
+        }.to_not change{ User.count }
+      end
+
+      it "should set a :notice flash" do
+        User.create(@user_with_email)
+        post :google_oauth2
+        expect(flash[:notice]).to be_present
+        expect(flash[:notice]).to eq "Успешно оторизиран чрез Google профил."
+      end
+
+      it "should update uid and provider" do
+        User.create(@user_with_email)
+        post :google_oauth2
+        expect(User.last.uid).to eql("12345")
+        expect(User.last.provider).to eql("google_oauth2")
+      end
+    end
+
+    context "post with invalid data" do
+      it "should not create user - invalid email" do
+        opts = {'info':{'email': '1234'}}
+        create_invalid_mock_hash(opts)
+        expect {
+          post :google_oauth2
+        }.to_not change{ User.count }
+
+      end
+
+      it "should not create user - missing uid" do
+        opts = {'uid': ""}
+        create_invalid_mock_hash(opts)
+        expect {
+          post :google_oauth2
+        }.to_not change{ User.count }
+
+        
```

Comment:
The whitespace is a kind of inconsistent around this file. In some places there are 2 empty lines. In other places there is 1 empty line. Can the white space be unified? Generally I believe the `it` and `context` blocks should be separated with one empty line. You can have a look at the convention followed in the other specs in the project.